loader/db.go

```diff
@@ -149,23 +149,44 @@ func (conn *Conn) executeSQL(ctx *tcontext.Context, queries []string, args ...[]
 	return err
 }
 
-func createConn(cfg *config.SubTaskConfig) (*Conn, error) {
-	dbDSN := fmt.Sprintf("%s:%s@tcp(%s:%d)/?charset=utf8mb4&maxAllowedPacket=%d",
-		cfg.To.User, cfg.To.Password, cfg.To.Host, cfg.To.Port, *cfg.To.MaxAllowedPacket)
-	baseConn, err := baseconn.NewBaseConn(dbDSN, &retry.FiniteRetryStrategy{}, baseconn.DefaultRawDBConfig())
-	if err != nil {
-		return nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeDownstream)
+// Close release db connection resource, return it to BaseDB.db connection pool
+func (conn *WorkerConn) Close() error {
+	if conn == nil || conn.baseConn == nil {
+		return nil
 	}
-
-	return &Conn{baseConn: baseConn, cfg: cfg}, nil
+	return conn.baseConn.Close()
 }
 
-func closeConn(conn *Conn) error {
-	if conn.baseConn == nil {
-		return nil
+func createConn(ctx context.Context, cfg *config.SubTaskConfig) (*conn.BaseDB, *WorkerConn, error) {
+	baseDB, err := conn.DefaultDBProvider.Apply(cfg.To)
+	if err != nil {
+		return nil, nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeDownstream)
 	}
+	baseConn, err := baseDB.GetBaseConn(ctx)
+	if err != nil {
+		return nil, nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeDownstream)
+	}
+	return baseDB, &WorkerConn{baseConn: baseConn, cfg: cfg}, nil
+}
 
-	return terror.DBErrorAdapt(conn.baseConn.Close(), terror.ErrDBDriverError)
+func createConns(tctx *tcontext.Context, cfg *config.SubTaskConfig, workerCount int) (*conn.BaseDB, []*WorkerConn, error) {
+	baseDB, err := conn.DefaultDBProvider.Apply(cfg.To)
+	if err != nil {
+		return nil, nil, terror.WithScope(terror.DBErrorAdapt(err, terror.ErrDBDriverError), terror.ScopeDownstream)
+	}
+	conns := make([]*WorkerConn, 0, workerCount)
+	for i := 0; i < workerCount; i++ {
+		baseConn, err := baseDB.GetBaseConn(tctx.Context())
+		if err != nil {
+			err := baseDB.Close()
```

Comment:
this `:` will cause returning `baseDB.Close()` in L185.

and should we need to close connections already in `conns`?

Reply:
Update logic [here](https://github.com/pingcap/dm/pull/266/files#diff-5f5bb90e4d87dad0cb25c377bb10679bR182)

if we close the baseDB, all connections generate by this baseDB will release.

Reply:
> if we close the baseDB, all connections generate by this baseDB will release.

can you show me the code in the database builtin pkg?

Reply:
After I check the code about `database/sql` and `MySQLDriver`, I found I made a mistake about `*sql.DB/Close`, It only close free connections and stop receiving new open connection requests, other connections won't release `net.Conn` resource, So, I put a member about [conns](https://github.com/pingcap/dm/pull/266/files#diff-184b4272aa0f082766a0d470548f0644R71)  in `BaseDB` which hold all connections generated from its `BaseDB`, then We can close all connections when we call `BaseDB.Close` function.